linkup-cli/src/start.rs

```diff
@@ -1,165 +1,231 @@
 use std::{
-    env, fs,
+    collections::HashMap,
+    fs,
+    io::stdout,
     path::{Path, PathBuf},
+    sync,
+    thread::{self, sleep, JoinHandle},
+    time::Duration,
 };
 
 use colored::Colorize;
+use crossterm::{cursor, ExecutableCommand};
 
 use crate::{
-    background_booting::{wait_for_dns_ok, BackgroundServices, LocalBackgroundServices},
     env_files::write_to_env_file,
-    linkup_file_path,
     local_config::{config_path, config_to_state, get_config},
-    paid_tunnel::{CfPaidTunnelManager, PaidTunnelManager},
-    services::tunnel::{CfTunnelManager, TunnelManager},
-    status::print_session_names,
-    system::{RealSystem, System},
-    LINKUP_LOCALDNS_INSTALL,
-};
-use crate::{
-    local_config::LocalState,
-    status::{server_status, ServerStatus},
-    CliError,
+    services::{self, BackgroundService},
 };
+use crate::{local_config::LocalState, CliError};
 
-pub fn start(config_arg: &Option<String>, no_tunnel: bool) -> Result<(), CliError> {
-    env_logger::init();
-    let is_paid = use_paid_tunnels();
-    let state = load_and_save_state(config_arg, no_tunnel, is_paid)?;
-    set_linkup_env(state.clone())?;
-    if is_paid {
-        start_paid_tunnel(
-            &RealSystem,
-            &CfPaidTunnelManager,
-            &LocalBackgroundServices,
-            &CfTunnelManager,
-            state,
-        )?;
-    } else {
-        start_free_tunnel(state, no_tunnel)?;
-    }
-    Ok(())
-}
+const LOADING_CHARS: [char; 10] = ['⠋', '⠙', '⠹', '⠸', '⠼', '⠴', '⠦', '⠧', '⠇', '⠏'];
 
-fn set_linkup_env(state: LocalState) -> Result<(), CliError> {
-    // Set env vars to linkup
-    for service in &state.services {
-        if let Some(d) = &service.directory {
-            set_service_env(d.clone(), state.linkup.config_path.clone())?
-        }
-    }
-    Ok(())
-}
-
-fn use_paid_tunnels() -> bool {
-    env::var("LINKUP_CLOUDFLARE_ACCOUNT_ID").is_ok()
-        && env::var("LINKUP_CLOUDFLARE_ZONE_ID").is_ok()
-        && env::var("LINKUP_CF_API_TOKEN").is_ok()
-}
-
-fn start_paid_tunnel(
-    sys: &dyn System,
-    paid_manager: &dyn PaidTunnelManager,
-    boot: &dyn BackgroundServices,
-    tunnel_manager: &dyn TunnelManager,
-    mut state: LocalState,
+/// # Arguments
+/// * `config_arg`  - Path to the Linkup config to be used as base in case `fresh_state` argument is `true`.
+/// * `no_tunnel`   - If there should not be a Cloudflare tunnel.
+/// * `fresh_state` - Boolean representing if should refresh the state to what is defined on `config_arg`.
+pub async fn start(
+    config_arg: &Option<String>,
+    no_tunnel: bool,
+    fresh_state: bool,
```

Comment:
These positionals were a bit confusing when I saw them elsewhere: it might be nice to have them as a struct instead?

Reply:
Changed in ec2fb792c624da2b3c3e8d99642ffc5db0d21250